File: src_local/utils/context_iterator.py
import logging
from pathlib import Path

logger = logging.getLogger(__name__)

class ContextIterator:
    """
    🧬 Iterador de Contexto PhD.
    Encapsula a lógica de travessia e filtragem de arquivos do projeto.
    """
    def __init__(self, project_root, context_data, integrity_guardian, ignored_files, stack):
        self.project_root = project_root
        self.context_data = context_data
        self.integrity_guardian = integrity_guardian
        self.ignored_files = ignored_files
        self.stack = stack

    def iter_auditable_files(self):
        """Gerador para iteração limpa e eficiente de arquivos auditáveis."""
        for file in self.context_data.keys():
            if self._should_audit_file(file):
                content = self.read_project_file(file)
                if content:
                    yield file, content

    def _should_audit_file(self, file):
        """Valida relevância via IntegrityGuardian."""
        if file in self.ignored_files: return False
        if self.context_data.get(file, {}).get("component_type") == "TEST": return False
        return self.integrity_guardian.is_relevant_file(file, self.stack)

    def read_project_file(self, rel_path):
        """Leitura segura."""
        abs_path = Path(self.project_root) / rel_path
        try:
            return abs_path.read_text(encoding='utf-8', errors='ignore')
        except Exception as e:
            logger.warning(f"⚠️ Falha na leitura do arquivo {rel_path}: {e}")
            return None

File: src_local/utils/context_iterator.ts
import winston from "winston";
import { Path } from "../core/path_utils.ts";

const logger = winston.child({ module: "ContextIterator" });

/**
 * 🔄 Iterador de Contexto PhD (Bun Version).
 */
export class ContextIterator {
    projectRoot: Path | null;
    map: Record<string, any>;
    guardian: any;
    ignoredFiles: string[];
    stack: string;

    constructor(projectRoot: string | null, contextMap: Record<string, any>, options: { integrityGuardian?: any; ignoredFiles?: string[]; stack?: string } = {}) {
        this.projectRoot = projectRoot ? new Path(projectRoot) : null;
        this.map = contextMap;
        this.guardian = options.integrityGuardian;
        this.ignoredFiles = options.ignoredFiles || [];
        this.stack = options.stack || "Universal";
        logger.debug(`🔄 [Iterator] Inicializado para ${this.stack} com ${Object.keys(contextMap).length} itens.`);
    }

    getPyFiles(): Record<string, any> {
        const pyFiles: Record<string, any> = {};
        for (const [p, d] of Object.entries(this.map)) {
            this.addIfPythonFile(p, d, pyFiles);
        }
        return pyFiles;
    }

    private addIfPythonFile(p: string, d: any, pyFiles: Record<string, any>) {
        if (p.endsWith('.py')) {
            pyFiles[p] = d;
        }
    }

    async *iterAuditableFiles(): AsyncGenerator<[string, string]> {
        for (const [relPath, metadata] of Object.entries(this.map)) {
            const content = await this.auditSingleFile(relPath, metadata);
            if (content) yield [relPath, content];
        }
    }

    private async auditSingleFile(relPath: string, metadata: any): Promise<string | null> {
        if (!this.shouldAudit(relPath, metadata)) return null;
        return await this.readFile(relPath);
    }

    private shouldAudit(relPath: string, metadata: any): boolean {
        const isIgnored = this.ignoredFiles.includes(relPath);
        const isTest = metadata.component_type === "TEST";
        return !isIgnored && !isTest;
    }

    private async readFile(relPath: string): Promise<string | null> {
        if (!this.projectRoot) return null;
        const absPath = this.projectRoot.join(relPath);

        try {
            return await this.getFileContent(absPath);
        } catch (e) {
            logger.warn(`⚠️ [Iterator] Falha ao ler ${relPath}: ${e}`);
            return null;
        }
    }

    private async getFileContent(path: Path): Promise<string | null> {
        if (await path.exists()) {
            return await Bun.file(path.toString()).text();
        }
        return null;
    }
}
